### PR TITLE
CDAP-658 common-authorization: authorization service and client

### PR DESCRIPTION
**Pull request by jwang47**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ like [CDAP](https://github.com/caskdata/cdap), [Coopr](https://github.com/caskda
 ## Modules
 
 The Cask Common Libraries consists of the following modules:
```
Comment (awholegunch) [inline README.md]: these modules:

File: README.md [hunk 2]
```diff
@@ -8,6 +8,7 @@ like [CDAP](https://github.com/caskdata/cdap), [Coopr](https://github.com/caskda
...
+* common-authorization: Authorization service and client for protecting resources using Access Control Lists (ACLs).
 * common-cli: CLI framework that provides a CLI with auto-completion and command-parsing.
 * common-http: simple HTTP utilities
```
Comment (awholegunch) [inline README.md]: Simple HTTP utilities

File: README.md [hunk 3]
```diff
@@ -8,6 +8,7 @@ like [CDAP](https://github.com/caskdata/cdap), [Coopr](https://github.com/caskda
 ## Modules
 
 The Cask Common Libraries consists of the following modules:
+* common-authorization: Authorization service and client for protecting resources using Access Control Lists (ACLs).
```
Comment (awholegunch) [inline README.md]: remove trailing periods (common)

File: common-authorization/README.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+AuthorizationService service = injector.getInstance(AuthorizationService.class);
+service.startAndWait();
+
+ListenableFuture<Service.State> completionFuture = Services.getCompletionFuture(service);
+completionFuture.get();
+```
+
+### Distributed Cluster Mode
+
+To run the `AuthorizationService` in a distributed cluster, you can do the following:
```
Comment (awholegunch) [inline common-authorization/README.md]: in a distributed cluster, use:

File: common-authorization/README.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+**Introduction**
+
+The Authorization module contains an authorization system that can be used to control access to protected resources.
+
+## Usage (Server)
+
+### In-Memory Mode
+
+To use authorization in your own project, first start the `AuthorizationService`:
```
Comment (awholegunch) [inline common-authorization/README.md]: do we need to show the includes?

File: common-authorization/README.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+);
+
+AuthorizationClient client = injector.getInstance(AuthorizationClient.class);
+```
+
+### Setting and Getting Access Control List (ACL) entries
+
+You can set and get ACL entries using the `ACLClient`:
+
+
```
Comment (awholegunch) [inline common-authorization/README.md:105]:
```

File: common-authorization/README.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+ObjectId secretFile = new ObjectId("FILE", "/tmp/sdf");
+SubjectId currentUser = SubjectId.ofUser("bob");
```
Comment (awholegunch) [inline common-authorization/README.md]: "Bob"

File: common-authorization/README.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+// grant "bob" WRITE access to the "/tmp/sdf" file
```
Comment (awholegunch) [inline common-authorization/README.md]: Grant "Bob"...

File: common-authorization/README.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+client.setACL(secretFile, currentUser, "WRITE");
+
+// get all ACL entries whose subject is "bob" and object is the "/tmp/sdf" file
```
Comment (awholegunch) [inline common-authorization/README.md]: Get all ACL entries whose subject is "Bob" and object is the "/tmp/sdf" file

File: common-authorization/README.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+### Controlling Access
+
+You can protect your resources by using the `ACLClient`:
+
+```
+AuthorizationClient client;
+ObjectId secretFile = new ObjectId("FILE", "/tmp/sdf");
+SubjectId currentUser = SubjectId.ofUser("bob");
+
+// verify that "bob" has WRITE access to the "/tmp/sdf" file
```
Comment (awholegunch) [inline common-authorization/README.md]: Verify that "Bob" has WRITE access to the "/tmp/sdf" file

File: common-authorization/README.md [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+// grant "bob" WRITE access to the "/tmp/sdf" file
+client.setACL(secretFile, currentUser, "WRITE");
+
+// get all ACL entries whose subject is "bob" and object is the "/tmp/sdf" file
+client.getACLs(secretFile, currentUser);
+```
+
+### Controlling Access
+
+You can protect your resources by using the `ACLClient`:
```
Comment (awholegunch) [inline common-authorization/README.md]: `AuthorizationClient`

File: common-authorization/README.md [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+```
+AuthorizationClient client;
+ObjectId secretFile = new ObjectId("FILE", "/tmp/sdf");
+SubjectId currentUser = SubjectId.ofUser("bob");
+
+// verify that "bob" has WRITE access to the "/tmp/sdf" file
+client.verifyAuthorized(secretFile, currentUser, "WRITE");
+```
+
+Alternatively, you can use `@RequiredPermissions` to guard access to method calls:
```
Comment (awholegunch) [inline common-authorization/README.md]: This tag does not match the code sample. Should it be "Requires"?

File: common-authorization/README.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+  new DiscoveryRuntimeModule().getInMemoryModules(),
+  new CGLibAuthorizationProxyRuntimeModule()
+);
+
+AuthorizationProxyFactory proxyFactory = injector.getInstance(AuthorizationProxyFactory.class);
+
+SecretFile secretFile = new SecretFile();
+SecretFile protectedSecretFile = proxyFactory.getProxy(secretFile, new ObjectId("FILE", "/tmp/sdf"));
+
+// will throw UnauthorizedException due to missing permissions
```
Comment (awholegunch) [inline common-authorization/README.md]: Will throw UnauthorizedException due to incorrect permissions

File: common-authorization/src/main/java/co/cask/common/authorization/ACLEntry.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.common.authorization;
+
+/**
+ * <p>
```
Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/ACLEntry.java]: The first sentence of a javadoc does not have paragraph tags around it. See 
http://www.liferay.com/community/wiki/-/wiki/Main/Javadoc+Guidelines#section-Javadoc+Guidelines-General+Guidelines

File: common-authorization/src/main/java/co/cask/common/authorization/ACLStore.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.common.authorization;
+
+import java.util.List;
+
+/**
+ * Provides ways to store and query {@link ACLEntry}s.
```
Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/ACLStore.java]: Provides methods for storing and querying {@link ACLEntry}s.

File: common-authorization/src/main/java/co/cask/common/authorization/ACLStore.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+ */
+public interface ACLStore {
+
+  /**
+   * Writes an {@link ACLEntry}.
+   *
+   * @param entry the {@link ACLEntry} to write
+   * @return true if entry had not yet existed
```
Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/ACLStore.java]: @return true if entry did not previously exist

File: common-authorization/src/main/java/co/cask/common/authorization/ACLStore.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+   * @return true if the {@link ACLEntry} exists
+   */
+  boolean exists(ACLEntry entry);
+
+  /**
+   * Fetches a list of {@link ACLEntry}s by {@code object} and {@code subject}.
+   *
+   * @param objectId the relevant object
+   * @param subjectId the relevant subject
+   * @return a list of {@link ACLEntry}s that have the {@code object} and {@code subject}.
```
Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/ACLStore.java]: @return the {@link ACLEntry}s with the {@code object} and {@code subject} [you don't give the return type in the description]

File: common-authorization/src/main/java/co/cask/common/authorization/AuthorizationContext.java [hunk 18]
```diff
@@ -0,0 +1,14 @@
+package co.cask.common.authorization;
```
Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/AuthorizationContext.java:16]: missing copyright statement

File: common-authorization/src/main/java/co/cask/common/authorization/DefaultAuthorizationContext.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright © 2012-2014 Cask Data, Inc.
```
Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/DefaultAuthorizationContext.java]: 2012 or 2014?

File: common-authorization/README.md (hunk 12 above)
Comment (awholegunch) [inline common-authorization/README.md]: Should this tag be "@RequiresPermission" (singular)? Or "@RequiredPermissions"? Just wondering...
"@RequiredPermissions" reads nicely...

File: common-authorization/src/main/java/co/cask/common/authorization/RequiresPermissions.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation for a method that requires permissions to execute.
+ *
+ * Proxy your annotated object with
+ * {@link co.cask.common.authorization.proxy.AuthorizationProxyFactory#getProxy(Object, ObjectId)}
+ * to have a proxy instance that verifies access control on methods annotated with {@link RequiresPermissions}.
```
Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/RequiresPermissions.java:30]: surround this sentence with paragraph tags

File: common-authorization/src/main/java/co/cask/common/authorization/UnauthorizedException.java [hunk 21]
```diff
@@ -0,0 +1,8 @@
+package co.cask.common.authorization;
```
Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/UnauthorizedException.java:16]: copyright block missing.

File: common-authorization/src/main/java/co/cask/common/authorization/UnauthorizedException.java [hunk 22]
```diff
@@ -0,0 +1,8 @@
...
+
+/**
+ * Thrown when a user tried to access an access-controlled resource without the necessary permissions.
```
Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/UnauthorizedException.java]: Thrown when a user tried to access a protected resource without the necessary permissions.

File: common-authorization/README.md (hunk 5 above)
Comment (jwang47) [inline common-authorization/README.md]: Should be fine without.

File: common-authorization/src/main/java/co/cask/common/authorization/UnauthorizedException.java (hunk 21 above)
Comment (jwang47) [inline common-authorization/src/main/java/co/cask/common/authorization/UnauthorizedException.java:16]: It's there, right?

Comment (awholegunch) [inline common-authorization/src/main/java/co/cask/common/authorization/UnauthorizedException.java:16]: Where was it hiding? It's sure there now...

File: README.md [hunk 23]
```diff
@@ -7,8 +7,9 @@ like [CDAP](https://github.com/caskdata/cdap), [Coopr](https://github.com/caskda
```
Comment (awholegunch) [inline README.md:7]: of Cask projects like -> of Cask projects such as

**Pull request closed without merging**